[wasm] Allow assemblies to be skipped for AOTMode=AotInterp

radical opened the pull request:
* Allow skipping AOT, per assembly with `%(_InternalForceInterpret)` metadata
* build: Rename `@(WasmAssembliesToBundle)` -> `@(WasmAssembly)`
* new `$(WasmNativeDebugSymbols)` to control symbols for dotnet.wasm (useful only for AOT case), which defaults to true

* Path handling fixes for AOT builds on windows
* add `create-packs.proj` for setting up packs from a local build, for quick local testing

radical committed:
[wasm] reduce visibility of `[AOT] /foo/bar.dll`  messages

radical committed:
[wasm] Allow skipping AOT, per assembly with %(InterpOnly) metadata

radical committed:
[wasm] build: Rename @(WasmAssembliesToBundle) -> @(WasmAssemblies)

- The reason for this change is that generating the app bundle is
  optional now. Eg. in case of blazor, which handles the layout itself.

- In that case, the assemblies will be used for other things during
  `WasmBuildApp` target (eg. AOT).
  - Change the name to reflect that

radical committed:
[wasm] Add tools/create-pack.proj for creating workload packs

pranavkm commented on eng/testing/tests.mobile.targets:
```diff
@@ -211,7 +211,7 @@
       </WasmSatelliteAssemblies>
 
-      <WasmAssembliesToBundle Include="$(PublishDir)\*.dll"/>
+      <WasmAssemblies Include="$(PublishDir)\*.dll"/>
```
Bit of a nit because I see this convention broken far too often, but it's common to use the singular case when naming items. So this would be

```suggestion
      <WasmAssembly Include="$(PublishDir)\*.dll"/>
```

pranavkm commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -77,12 +80,13 @@
     </ItemGroup>
     <ItemGroup>
-      <_AotInputAssemblies Include="@(_WasmAssemblies->Distinct())">
+      <_AotInputAssemblies Include="@(_WasmAssembliesInternal->Distinct())" Condition="'%(_WasmAssembliesInternal.InterpOnly)' != 'true'">
```
`InterpOnly`  

* Could we use a non-abbreviated name `InterprettedOnly`? Users might need to configure it. 
* Might also want to consider something like `AOTMode=Intepretted`. 

This would look like:

```xml
_AotInputAssemblies 
   Include="@(_WasmAssembliesInternal->Distinct())" 
   Condition="'%(_WasmAssembliesInternal.AOTMode)' != 'Interpretted'">
```

radical replied:
> Might also want to consider something like AOTMode=Intepretted.

Does it make sense to have the possible values for `AOTMode` to be as what Android, and iOS have? @lewing @marek-safar @jonathanpeppers

jonathanpeppers replied:
What we have now:

https://github.com/xamarin/xamarin-android/blob/accc846e35ccc2b21e2a472d51ab58ed0682643a/src/Xamarin.Android.Build.Tasks/Tasks/Aot.cs#L20-L27

It kind of maps to the native side:

https://github.com/dotnet/runtime/blob/1d9e50cb4735df46d3de0cee5791e97295eaf588/src/mono/mono/mini/jit.h#L46-L72

jonathanpeppers replied:
I don't think our integer values matter at build time for what is passed at runtime.

radical replied:
And those c# enum names are the strings used with `$(AOTMode)` (or equivalent property name)?

jonathanpeppers replied:
Our property is `$(AndroidAotMode)` right now, but we can use a new one in .NET 6 to match other platforms. I'm fine with `$(AOTMode)` if that is what we pick.

radical replied:
Sorry, I'm not being clear. What values do you use for `$(AndroidAotMode)`? Are they `$(AndroidAotMode)=normal`, `$(AndroidAotMode)=hybrid` etc, to match the C# enum?

jonathanpeppers replied:
It looks like they all map exactly to this C# enum, except for `Intepreter`:

https://github.com/xamarin/xamarin-android/blob/d8df924702a753bbba8d317b3d5e88dde10f7a52/src/Xamarin.Android.Build.Tasks/Tasks/Aot.cs#L96-L122

radical committed:
Merge remote-tracking branch 'origin/master' into aot-interp

radical committed:
[wasm] Rename @(WasmAssemblies) -> @(WasmAssembly)

steveisok commented on src/mono/wasm/tools/create-pack.proj:
```diff
@@ -0,0 +1,74 @@
+<Project>
```
@radical is your intent to do away with some of https://github.com/dotnet/runtime/tree/master/src/mono/nuget ?

radical replied:
This is just for quick local setup for testing.

radical committed:
[wasm] don't copy dotnet.* to appbundle unnecessarily

This was added earlier to support blazor workload which only uses AOT.
But that has since been fixed to correctly use
`WasmBuildApp`+`@(WasmNativeAsset)`, so this can be removed.

radical committed:
fix md linter error

radical committed:
[wasm] %(WasmAssembly.InterpOnly) -> %(WasmAssembly.InterpreterOnly)

radical committed:
[wasm] quote file paths in shell invocations

radical committed:
Update src/mono/wasm/build/WasmApp.targets

Co-authored-by: Pranav K <prkrishn@hotmail.com>

radical committed:
[wasm] Quote assembly path when invoking mono-aot-cross

radical committed:
[wasm] avoid trailing windows slash escaping the ending double quote

radical committed:
[wasm] Use the msbuild equivalent of Path.Combine in the targets/props

.. and enable it by default.

radical committed:
[wasm] WasmApp.targets: Add $(WasmNativeDebugSymbols) to build dotnet.wasm with symbols

.. defaults to True.

radical committed:
[wasm] Build dotnet.wasm with symbols, even in Release if NOSTRIP is not set

radical committed:
Merge remote-tracking branch 'origin/master' into quote-paths

radical committed:
[wasm] Run the build native step for %(InterpreterOnly) assemblies too

radical committed:
[wasm] rename %(InterpreterOnly) -> %(_Unsupported_InterpreterOnly_)

radical committed:
[wasm] Fix test failures caused by incorrect checks

1. tests in `src/tests` failed with:

`Bug: The wasm build started with 2 assemblies, but completed with 11.`

.. this was because they build with `$(WasmResolveAssembliesBeforeBuild)=true`
and that adds to the original list of assemblie. So, skip checking in
that case.

2. AOT functional tests failed with:

`Bug: The wasm build started with 356 assemblies, but completed with 178. `

.. this was because the original list had dupes! So, ensure to count
only distinct items.

radical committed:
[wasm] `%(_Unsupported_InterpreterOnly_) -> %(_InternalForceInterpret)`

radical committed:
[wasm] dedupe @(WasmAssembly)

radical commented:
Unrelated test failures in `Libraries Test Run release {coreclr,mono} Linux x64 Debug`

radical commented:
Test failures being tracked in https://github.com/dotnet/runtime/issues/48751

lewing reviewed (commented):
The changes look good to me.  I'd like to come to some broader consensus on naming, especially Aot vs AOT and the (Aot|AOT)Mode bits.

lewing commented:
But I'm ok to approve to unblock things if @pranavkm is waiting on this?

radical committed:
Merge remote-tracking branch 'origin/master' into quote-paths

radical committed:
Merge remote-tracking branch 'origin/main' into quote-paths

lewing committed:
Merge branch 'main' into aot-build

radical committed:
[wasm] Check that EMSDK_PATH is set for WasmBuildNative

radical committed:
wasm] Disable il stripping completely

This uses `mono-cil-strip` from a mono installation. And in it's current
form it can cause issues, so disabling it for now.

radical commented:
The parts of this that didn't do naming changes have been merged as part of #49446 . I will open a new PR, when we have a consensus on the naming.
